### PR TITLE
PRO X Wireless Headset

### PR DESCRIPTION
**Pull request by jaypikay**
Logitech devices seem a bit strange when reporting their voltage. As far as I could see the PRO X reports the voltage capacity.

- Battery status
  - charging
  - discharging (returns voltage)
- Set inactivity time

### CODE REVIEW

File: src/devices/logitech_gpro.c [hunk 1]
```diff
@@ -2,8 +2,14 @@
 #include "../utility.h"
 #include "logitech.h"
 
+#include <unistd.h>
+
+#include <math.h>
 #include <string.h>
 
+#define MSG_SIZE    20
```
Comment (Sapd) [inline src/devices/logitech_gpro.c:10]: The MSG_SIZE is not used anywhere, so we can probably remove it

**Comment by jaypikay** [comment]
Just noticed that #131 is related.

**Comment by jaypikay** [comment]
Just noticed some quirks with the voltage. When it gets lower the percentage isn't correct anymore

**Comment by Sapd** [comment]
Yes, you used a linear formular for the voltage. However LI-Ions usually hold their voltage for a long time (slowley decreasing) and then suddenly decreasing when they are nearing under 20%.

E.g. here are examples: https://hobbygraderc.com/lipo-voltage-chart/

You can use a polynomial regression calculator to make a mathematical curve: https://arachnoid.com/polysolve/ (the website will even generate C code)

Or you can use simple splines:
Simply set p and v to your liking, which comes nearest to your headset
```
static float estimate_battery_level(uint16_t voltage) {
    int p[] = { 100, 60, 10, 0 };
    int v[] = { 4200, 3870, 3690, 3270 };
    int size = 4;
    
    float percent = 0;
    for(int i = 0; i < size; ++i) {
        // if >= then 100%
        if(voltage >= v[i]) {
            percent = p[i];
            break;
        } 
        
        // if not last
        if(i < size-1 && voltage >= v[i+1]) {
            percent = p[i+1] + (voltage-v[i+1])/(v[i]-v[i+1])*(p[i]-p[i+1]);
            break;
        }
       return percent;
}
 ```

We can put the function in utility.h with p and v as parameters, so every headset can use it if it wants to use that method.

**Comment by jaypikay** [comment]
Hey thank you, for you post.

1. Inactivity timeout: I don't know, haven't looked into this, but it's an idea. I can try to look into it.
2. Need to look into it.
3. Does not look like this could be a problem, the head set manages the standby by itself.
4. There shouldn't be any differences. Except the microphone stuff. Logitech uses this *Blue* thing stuff, but it should be possible to achieve similar microphone quality using different audio filter tools, for Pulseaudio/Pipewire or Alsa and Jack. I do not have more details on this topic, as I do not care.
5. The "forks" are metal

**Comment by Bananaman** [comment]
@jaypikay Thank you so much for your answers. I apologize, I thought I had replied earlier. I am really grateful that you took the time.

I'm glad to hear that the automatic standby works on Linux too, so the battery doesn't go empty constantly.

And you're right, Logitech uses some Blue plugins on Windows. But I discovered that we can get really good audio quality on Linux too by using the EasyEffects app: https://github.com/wwmm/easyeffects

Thanks for everything. You're a fantastic person. :)

**Pull request merged**